cloudinit/ssh_util.py

```diff
@@ -683,3 +701,116 @@ def get_opensshd_upstream_version():
+
+
+def _get_early_key_fifo_path(rundir: str) -> pathlib.Path:
+    return pathlib.Path(rundir, "ssh-keygen-finished")
+
+
+def _write_and_close(path: pathlib.Path, data: bytes) -> None:
+    path.write_bytes(data)
+    path.unlink()
+
+
+def _early_generate_host_keys_body(
+    rundir: str, early_key_fifo_path: pathlib.Path
+) -> None:
+    key_dir = get_early_host_key_dir(rundir)
+    key_dir.mkdir(mode=0o600, exist_ok=False)
+
+    for key_type in GENERATE_KEY_NAMES:
+        path = key_dir / (KEY_NAME_TPL % key_type)
+        stdout_path = path.with_suffix(".stdout")
+        stderr_path = path.with_suffix(".stderr")
+        processes = []
+        with open(stdout_path, "w") as stdout, open(
+            stderr_path, "w"
+        ) as stderr:
+            try:
+                # Using subprocess.Popen instead of subp.subp to run
+                # multiple ssh-keygen commands in parallel.
+                p = subprocess.Popen(
+                    [
+                        "ssh-keygen",
+                        "-t",
+                        key_type,
+                        "-N",
+                        "",
+                        "-f",
+                        path,
+                    ],
+                    stdout=stdout,
+                    stderr=stderr,
+                )
+                processes.append(p)
+            except Exception as e:
+                LOG.warning("Failed to generate %s host key: %s", key_type, e)
+    for process in processes:
+        if process.wait() != 0:
+            _write_and_close(early_key_fifo_path, b"failed")
+            return
+    _write_and_close(early_key_fifo_path, b"done")
+
+
+def _early_generate_host_keys(
+    rundir: str, early_key_fifo_path: pathlib.Path
+) -> None:
+    try:
+        _early_generate_host_keys_body(rundir, early_key_fifo_path)
+    except Exception as e:
+        LOG.warning("Failed to generate host keys: %s", e)
+        _write_and_close(early_key_fifo_path, b"failed")
+
+
+def start_early_generate_host_keys(rundir: str):
+    if all(
+        pathlib.Path(KEY_FILE_TPL % key).exists() for key in GENERATE_KEY_NAMES
+    ):
+        LOG.debug(
+            "Existing host keys present; skipping early host key generation"
+        )
+        return
+    early_key_fifo_path = _get_early_key_fifo_path(rundir)
+    early_key_fifo_path.parent.mkdir(mode=0o700, exist_ok=True)
+    os.mkfifo(early_key_fifo_path)
+    try:
+        Process(
+            target=_early_generate_host_keys,
+            args=(rundir, early_key_fifo_path),
+            daemon=True,
+        ).start()
+    except Exception as e:
+        LOG.warning("Failed to start early host key generation: %s", e)
+        early_key_fifo_path.unlink()
+
+
+def wait_for_early_generated_keys(rundir: str) -> List[KeyPair]:
+    early_key_fifo_path = _get_early_key_fifo_path(rundir)
+    if not early_key_fifo_path.exists():
+        return []
+    if early_key_fifo_path.read_bytes() != b"done":
```

Comment:
If not complete, this will block. It would be good to know if this happens, such as by using `performance.Timed` or `performance.timed()`